keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,211 @@
+
+This KEP started life as [feature #2273](https://github.com/kubernetes/community/pull/2273).  Please follow that link for historical context.
+
+
+## Motivation
+Implementing support for embedding volumes directly in pod specs would allow external CSI drivers to have feature parity with their in-tree counter parts.  There are several reasons why this is desirable:
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
+* This feature would make it possible to create new types of CSI drivers such as ephemeral volume drivers.  They can be used to inject arbitrary states such as configuration, secrets, identy, variables or similar information inside pods. 
+* This can also provide a migration path for older Flex-style drivers (allowing the deprecation of Flex.)
+
+### Goals 
+* Provide a high level design for inline CSI volumes support
+* Define API changes needed to support this feature
+* Inline CSI volumes should work with existing CSI drivers
+* Inline CSI volumes should support ephemeral-style volumes
+* Ensure that inline CSI volumes usage is secure
+
+### Non-goals
+* This feature does not introduce nor require new CSI spec features
+* This feature should not force existing CSI drivers to change
+
+## User stories
+* As a storage provider, I want to be able to create CSI drivers that support persistent volumes that can be nested within pod specs.  These volumes would work similarly to how my current in-tree drivers work.
+* As a storage provider, I want to use the CSI API to develop arbitrary drivers that can mount ephemeral volumes which follow the lifecycles of pods where they are embedded.   These ephemeral volumes do not follow the full CSI storage workflow (of provision, stage, attach, mount, unmount, detach, unstage, delete)
+* As a user, I want to be able to deploy pods, with CSI volumes embedded in their specs, without the use of PV/PVC.  
+  * An administrator would create the volume ahead of time with a volume handle.  
+  * User can then setup pods with inline CSI volumes using reference of existing handle.
+
+* As a user I want to be able to define pod specs with embedded ephemeral CSI volumes that are created/mounted when the pod is deployed and are deleted when the pod goes away.
+
+### Examples
+
+**Example 1**
+A pod spec with an inline CSI volume.  Note that `volumeHandle` refers to a pre-existing volume.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testpod
+spec:
+  containers:
+    ...
+  volumes:
+      - name: vol
+        csi:
+          lifecycle: persistent
+          driver: io.kubernetes.storage.mock
+          volumeAttributes:
+              name: "Mock Volume 1"
```

Comment:
Would be great if we could support type `map[string][]string` or `map[string]map[string]string{}` instead of `map[string]string`. It would make this more flexible. We had this issue with `options` as part of the flex volume spec. It was very limiting. https://github.com/Azure/kubernetes-keyvault-flexvol/pull/49

Reply:
How would that map? csi volumeAttributes only support map[string]string I believe.

Reply:
(in the csi specification I mean.)

Reply:
yes currently. this comment is an enhancement ask. we ran into this limitation with flexvolume when we needed one of the attributes to be an array.

Reply:
Unfortunately these attributes have to stay flat to stay compatible with CSI.